<center><b>TOÁN TỬ TRONG PYTHON</b><center>

Toán tử được dùng để thực hiện các phép toán trên các biến và giá trị.

Python chia các toán tử thành các nhóm sau:
- Toán tử toán học (Arithmetic operators)
- Toán tử gán (Assignment operators)
- Toán tử so sánh (Comparison operators)
- Toán tử logic (Logical operators)
- Toán tử xác định (Identity operators)
- Toán tử quan hệ (Membership operators)
- Toán tử thao tác với bit (Bitwise operators)

**1.Toán tử toán học** 
<br>Các toán tử toán học được dùng với các giá trị là số để thực hiện các phép toán toán học:<br>
<table style='text-align:center;width:70%;margin:0 10% 0 10%;border:1px solid black'>
    <tr style='background:green;'>
        <th>Ký hiệu</th>
        <th>Tên toán tử</th>
        <th>Biểu thức</th>
        <th>Ví dụ</th>
    </tr>
    <tr style='background: #3399FF; color:blue'>
        <td>+</td>
        <td>Cộng</td>
        <td>x + y</td>
        <td> 3 + 4 = 7</td>
    </tr>
    <tr style='background: lightblue; color:blue'>
        <td>-</td>
        <td>Hiệu</td>
        <td>x - y</td>
        <td> 5 - 3 = 2</td>
    </tr>
    <tr style='background: #3399FF; color:blue'>
        <td>*</td>
        <td>Tích</td>
        <td>x * y</td>
        <td>3 * 4 = 12</td>
    </tr>
    <tr style='background: lightblue; color:blue'>
        <td>/</td>
        <td>Chia</td>
        <td>x / y</td>
        <td>5 / 2 = 2.5</td>
    </tr>
    <tr style='background: #3399FF; color:blue'>
        <td>%</td>
        <td>Chia lấy dư</td>
        <td>x % y</td>
        <td>5 % 2 = 1</td>
    </tr>
    <tr style='background: lightblue; color:blue'>
        <td>**</td>
        <td>Mũ</td>
        <td>x ** y</td>
        <td>5 ** 2 = 25</td>
    </tr>
    <tr style='background: #3399FF; color:blue'>
        <td>//</td>
        <td>Chia lấy nguyên</td>
        <td>x // y</td>
        <td>5 // 2 = 2</td>
    </tr>
</table>

**2. Toán tử gán**
<br>Được dùng để giá trị vào biến

<table style='text-align:center;width:70%;margin:0 10% 0 10%;'>
    <tr style='background:orange;color:white;'>
        <th>Toán tử</th>
        <th>Ví dụ</th>
        <th>Biểu thức tương đương</th>
    </tr>
    <tr style='background: #FFFF66; color:blue;'>
        <td>=</td>
        <td>x = 5</td>
        <td>x = 5</td>
    </tr>
    <tr style='background: #FFFF99; color:blue;'>
        <td>+=</td>
        <td>x += 5</td>
        <td>x = x + 5</td>
    </tr>
    <tr style='background: #FFFF66; color:blue;'>
        <td>-=</td>
        <td>x -= 5</td>
        <td>x = x - 5</td>
    </tr>
    <tr style='background: #FFFF99; color:blue;'>
        <td>*=</td>
        <td>x *= 5</td>
        <td>x = x * 5</td>
    </tr>
    <tr style='background: #FFFF66; color:blue;'>
        <td>/=</td>
        <td>x /= 5</td>
        <td>x = x / 5</td>
    </tr>
    <tr style='background: #FFFF99; color:blue;'>
        <td>%=</td>
        <td>x %= 5</td>
        <td>x = x % 5</td>
    </tr>
    <tr style='background: #FFFF66; color:blue;'>
        <td>//=</td>
        <td>x //= 5</td>
        <td>x = x // 5</td>
    </tr>
    <tr style='background: #FFFF99; color:blue;'>
        <td>**=</td>
        <td>x **= 5</td>
        <td>x = x ** 5</td>
    </tr>
    <tr style='background: #FFFF66; color:blue;'>
        <td>&=</td>
        <td>x &= 5</td>
        <td>x = x & 5</td>
    </tr>
    <tr style='background: #FFFF99; color:blue;'>
        <td>|=</td>
        <td>x |= 5</td>
        <td>x = x | 5</td>
    </tr>
    <tr style='background: #FFFF66; color:blue;'>
        <td>^=</td>
        <td>x ^= 5</td>
        <td>x = x ^ 5</td>
    </tr>
    <tr style='background: #FFFF99; color:blue;'>
        <td>>>=</td>
        <td>x >>= 5</td>
        <td>x = x >> 5</td>
    </tr>
    <tr style='background: #FFFF66; color:blue;'>
        <td><<=</td>
        <td>x <<= 5</td>
        <td>x = x << 5</td>
    </tr>
</table>

In [ ]:
x = 3
x += 5 #Tương tự x = x + 5 = 3 + 5 = 8
x **= 5 #Tương tự x = x**5 = 3**5 = 243

**3. Toán tử so sánh**
<br>Được dùng để so sánh hai giá trị:<br>
<table style='text-align:center;width:70%;margin:0 10% 0 10%;'>
    <tr style='background: #FF3333; color:white;'>
        <th>Ký hiệu</th>
        <th>Tên toán tử</th>
        <th>Biểu thức</th>
        <th>Ví dụ</th>
        <th>Kết quả</th>
    </tr>
    <tr style='background: #FF9999; color:white;'>
        <td>==</td>
        <td>So sánh bằng</td>
        <td>x == y</td>
        <td>3 == 4</td>
        <td>False</td>
    </tr>
    <tr style='background: #FF6666; color:white;'>
        <td>!=</td>
        <td>So sánh không bằng</td>
        <td>x != y</td>
        <td> 2 != 3</td>
        <td>True</td>
    </tr>
    <tr style='background: #FF9999; color:white;'>
        <td>></td>
        <td>So sánh lớn hơn</td>
        <td>x > y</td>
        <td>3 > 4</td>
        <td>False</td>
    </tr>
    <tr style='background: #FF6666; color:white;'>
        <td><</td>
        <td>So sánh bé hơn</td>
        <td>x < y</td>
        <td>5 < 2</td>
        <td>False</td>
    </tr>
    <tr style='background: #FF9999; color:white;'>
        <td>>=</td>
        <td>So sánh lớn hơn hoặc bằng</td>
        <td>x >= y</td>
        <td>5 >= 2</td>
        <td>True</td>
    </tr>
    <tr style='background: #FF6666; color:white;'>
        <td><=</td>
        <td>So sánh bé hơn hoặc bằng</td>
        <td>x <= y</td>
        <td>5 <= 2</td>
        <td>False</td>
    </tr>
</table>

**4. Toán tử logic**
<br>Dùng để kết hợp nhiều câu lệnh điều kiện lại với nhau: (x = 4)

<table style='text-align:center;width:70%;margin:0 10% 0 10%;'>
    <tr style='background:#00FF00;color:blue'>
        <th>Toán tử</th>
        <th>Mô tả</th>
        <th>Ví dụ</th>
        <th>Kết quả</th>
    </tr>
    <tr style='background:#66FF66;color:blue'>
        <td>and</td>
        <td>Trả về True nếu tất cả các câu lệnh điều kiện đều là True</td>
        <td>x < 5 and x < 10</td>
        <td>True</td>
    </tr>
    <tr style='background:#99FF99;color:blue'>
        <td>or</td>
        <td>Trả về True nếu một trong các câu lệnh điều kiện là True</td>
        <td>x < 5 or x < 4</td>
        <td>True</td>
    </tr>
    <tr style='background:#66FF66;color:blue'>
        <td>not</td>
        <td>Đảo ngược kết quả, trả về False nếu kết quả là True</td>
        <td>not(x < 5 and x < 10)</td>
        <td>False</td>
    </tr>
</table>

**5. Toán tử xác định**
<br>Dùng để so sánh các object, không phải so sánh các object bằng nhau, mà là so sánh xem cac object có cùng là một object hay cùng vị trí bộ nhớ hay không:

<table style='text-align:center;width:70%;margin:0 10% 0 10%;'>
    <tr style='background:#FF007F;color:white'>
        <th>Toán tử</th>
        <th>Mô tả</th>
        <th>Ví dụ</th>
    </tr>
    <tr style='background:#FF66FF;color:white'>
        <td>is</td>
        <td>Trả về True nếu cả 2 biến <b style='color:blue'>cùng</b> một object</td>
        <td>x is y</td>
    </tr>
    <tr style='background:#FF99FF;color:white'>
        <td>is not</td>
        <td>Trả về True nếu cả 2 biến <b style='color:blue'>không</b> cùng một object</td>
        <td>x is not y</td>
    </tr>
</table>

In [2]:
x = ["apple", "banana", "orange"]
y = ["apple", "banana", "orange"]
z = x
print(x is z)
print(x is y)
print(x == y)
print(x is not y)

True
False
True
True


**6. Toán tử quan hệ**
<br>Dùng để kiểm tra xem một object <b style='color:red'>a</b> có trong một object <b style='color:red'>b</b> hay không: 

<table style='text-align:center;width:70%;margin:0 10% 0 10%;'>
    <tr style='background:#FFFF00;color:black'>
        <th>Toán tử</th>
        <th>Mô tả</th>
        <th>Ví dụ</th>
    </tr>
    <tr style='background:#FFFF66;color:black'>
        <td>ns</td>
        <td>Trả về True nếu object <b style='color:red'>a</b> <b style='color:blue'>có trong</b> <b style='color:red'>b</b></td>
        <td>a is b</td>
    </tr>
    <tr style='background:#FFFF99;color:black'>
        <td>not in</td>
        <td>Trả về True nếu object <b style='color:red'>a</b> <b style='color:blue'>không tồn tại</b> trong object <b style='color:red'>b</b></td>
        <td>a is not b</td>
    </tr>
</table>

In [5]:
x = ["apple", "banana"]

print("banana" in x)
print(2 in x)
print(True not in x)

True
False
True


**7. Toán tử thao tác với bit**
<br>Dùng để so sánh các số ở dạng nhị phân

<div style='text-align:center;width:70%;margin:0 10% 0 10%;column-count:2'>
    <div style='float:left;'>
        <table>
            <tr>
                <th>Số thập phân</th>
                <th>Số nhị phân</th>
            </tr>
            <tr>
                <td>0</td>
                <td>0000 0000</td>
            </tr>
            <tr>
                <td>1</td>
                <td>0000 0001</td>
            </tr>
            <tr>
                <td>2</td>
                <td>0000 0010</td>
            </tr>
            <tr>
                <td>3</td>
                <td>0000 0011</td>
            </tr>
            <tr>
                <td>4</td>
                <td>0000 0100</td>
            </tr>
            <tr>
                <td>5</td>
                <td>0000 0101</td>
            </tr>
            <tr>
                <td>6</td>
                <td>0000 0110</td>
            </tr>
            <tr>
                <td>7</td>
                <td>0000 0111</td>
            </tr>
        </table>
    </div>
    <div style='float:right;'>
        <table>
            <tr>
                <th>Số thập phân</th>
                <th>Số nhị phân</th>
            </tr>
            <tr>
                <td>8</td>
                <td>0000 1000</td>
            </tr>
            <tr>
                <td>9</td>
                <td>0000 1001</td>
            </tr>
            <tr>
                <td>10</td>
                <td>0000 1010</td>
            </tr>
            <tr>
                <td>11</td>
                <td>0000 1011</td>
            </tr>
            <tr>
                <td>12</td>
                <td>0000 1100</td>
            </tr>
            <tr>
                <td>13</td>
                <td>0000 1101</td>
            </tr>
            <tr>
                <td>14</td>
                <td>0000 1110</td>
            </tr>
            <tr>
                <td>15</td>
                <td>0000 1111</td>
            </tr>
        </table>
    </div>
</div>

<table style='text-align:center;width:80%;margin:0 10% 0 10%;'>
    <tr style='background:#00CC00;color:white;'>
        <th>Toán tử</th>
        <th>Tên toán tử</th>
        <th>Mô tả</th>
        <th>Ví dụ</th>
    </tr>
    <tr style='background:#66FF66;color:white;'>
        <td>&</td>
        <td>AND</td>
        <td>Nếu cả 2 bit đều là 1 thì giá trị là 1</td>
        <td>x & y</td>
    </tr>
    <tr style='background:#33FF33;color:white;'>
        <td>|</td>
        <td>OR</td>
        <td>Nếu 1 trong 2 bit là 1 thì giá trị là 1</td>
        <td>x | y</td>
    </tr>
    <tr style='background:#66FF66;color:white;'>
        <td>^</td>
        <td>XOR</td>
        <td>Nếu chỉ có 1 trong 2 bit là 1 thì giá trị là 1</td>
        <td>x ^ y</td>
    </tr>
    <tr style='background:#33FF33;color:white;'>
        <td>~</td>
        <td>NOT</td>
        <td>Nếu bit là 1 thì giá trị là 0, và ngược lại</td>
        <td>x ~ y</td>
    </tr>
    <tr style='background:#66FF66;color:white;'>
        <td><<</td>
        <td>Dịch bit sang trái</td>
        <td>Dịch các bit sang trái bằng cách đặt bit 0 vào bên phải của dãy bit và loại bỏ bit ngoài cùng bên trái</td>
        <td>x << y</td>
    </tr>
    <tr style='background:#33FF33;color:white;'>
        <td>>></td>
        <td>Dịch bit sang phải</td>
        <td>Dịch các bit sang phải bằng cách đặt bản sao của bit ngoài cùng bên trái vào bên trái của dãy bit và loại bỏ bit ngoài cùng bên phải</td>
        <td>x >> y</td>
    </tr>
</table>

**Cách chuyển đổi giá trị thập phân sang nhị phân theo bảng sau:**

<table style='text-align:center;width:80%;margin:0 10% 0 10%;'>
    <tr>
        <td>Index</td>
        <td>7</td>
        <td>6</td>
        <td>5</td>
        <td>4</td>
        <td>3</td>
        <td>2</td>
        <td>1</td>
        <td>0</td>
    </tr>
    <tr style='background:#0078d7; color: white;'>
        <td>Tổng bit ở thập phân</td>
        <td>128</td>
        <td>64</td>
        <td>32</td>
        <td>16</td>
        <td>8</td>
        <td>4</td>
        <td>2</td>
        <td>1</td>
    </tr>
    <tr>
        <td>Giá trị bit</td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
        <td></td>
    </tr>
</table>

**8. Độ ưu tiên của các toán tử**
<br>Dùng để mô tả thứ tự thực hiện các phép toán

Ví dụ 1: Dấu ngoặc đơn () có thứ tự ưu tiên cao nhất, điều này có nghĩa là biểu thức trong dấu ngoặc () phải được đánh giá đầu tiên:

In [1]:
print((6 + 3) - (6 + 3))

0


Ví dụ 2: Phép nhân * có độ ưu tiên cao hơn phép cộng +, do đó phép nhân sẽ được đánh giá trước phép cộng:

In [2]:
print(100 + 5*3)

115


Thứ tự ưu tiên được mô tả trong bảng sau, bắt đầu với độ ưu tiên cao nhất (từ trên xuống):

<table style='text-align:center;width:80%;margin:0 10% 0 10%;'>
    <tr style='background:#0000CC'>
        <th>Toán tử</th>
        <th>Mô tả</th>
        <th>Ví dụ</th>
        <th>Kết quả</th>
    </tr>
    <tr style='background:#9999FF;color:blue;'>
        <td>()</td>
        <td>Dấu ngoặc đơn</td>
        <td>(6 + 3) - 2</td>
        <td>7</td>
    </tr>
    <tr style='background:#6666FF;color:white;'>
        <td>**</td>
        <td>Toán tử mũ</td>
        <td>3 ** 3 - 2</td>
        <td>25</td>
    </tr>
    <tr style='background:#9999FF;color:blue;'>
        <td>+x -x ~x</td>
        <td>Toán tử cộng 1 ngôi, trừ 1 ngôi, toán tử NOT</td>
        <td>100 + ~3</td>
        <td>96</td>
    </tr>
    <tr style='background:#6666FF;color:white;'>
        <td>* / // %</td>
        <td> Nhân, chia, chia lấy nguyên, chia lấy dư</td>
        <td>100 + 5 * 3</td>
        <td>115</td>
    </tr>
    <tr style='background:#9999FF;color:blue;'>
        <td>+ -</td>
        <td> Cộng, trừ</td>
        <td>100 - 5 * 3</td>
        <td>85</td>
    </tr>
    <tr style='background:#6666FF;color:white;'>
        <td><< >></td>
        <td> Dịch bit trái, dịch bit phải</td>
        <td>8 >> 4 - 2</td>
        <td>2</td>
    </tr>
    <tr style='background:#9999FF;color:blue;'>
        <td>&</td>
        <td>Bitwise AND</td>
        <td>6 & 2 + 1</td>
        <td>2</td>
    </tr>
    <tr style='background:#6666FF;color:white;'>
        <td>^</td>
        <td>Bitwise XOR</td>
        <td>6 ^ 2 + 1</td>
        <td>5</td>
    </tr>
    <tr style='background:#9999FF;color:blue;'>
        <td>|</td>
        <td>Bitwise OR</td>
        <td>6 | 2 + 1</td>
        <td>7</td>
    </tr>
    <tr style='background:#6666FF;color:white;'>
        <td>== != > >= < <= is is not in not in</td>
        <td>Toán tử so sánh, xác định, quan hệ</td>
        <td>5 == 4 + 1</td>
        <td>True</td>
    </tr>
    <tr style='background:#9999FF;color:blue;'>
        <td>not</td>
        <td>Logical NOT</td>
        <td>not 5 == 5</td>
        <td>False</td>
    </tr>
    <tr style='background:#6666FF;color:white;'>
        <td>and</td>
        <td>Logical AND</td>
        <td>1 or 2 and 3</td>
        <td>1</td>
    </tr>
    <tr style='background:#9999FF;color:blue;'>
        <td>or</td>
        <td>Logical OR</td>
        <td>4 or 5 + 10 or 8</td>
        <td>4</td>
    </tr>
</table>

<i>Nếu hai toán tử có cùng độ ưu tiên thì biểu thức được đánh giá từ trái sang phải</i>

In [3]:
print(5 + 4 - 7 + 3)

5


<footer style='margin: 50px 50px 0px 0px;'>
    <div>
    <hr/>
    Biên dịch: Nguyễn Ngọc Dương 
    <br>Facebook: <a href="https://www.facebook.com/danh.kean.353803/">https://www.facebook.com/danh.kean.353803/</a>
    <br>Email: <a>ngocduong5642@gmail.com</a>
    <div>
</footer>